public/app/features/dashboard-scene/settings/variables/components/QueryVariableForm.test.tsx

```diff
@@ -0,0 +1,248 @@
+  };
+
+  function setup(props?: React.ComponentProps<typeof QueryVariableEditorForm>) {
+    return render(<QueryVariableEditorForm {...defaultProps} {...props} />);
+  }
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should render the component with initializing the components correctly', () => {
+    const { getByTestId } = setup();
+    const dataSourcePicker = getByTestId(selectors.components.DataSourcePicker.container);
+    //const queryEditor = getByTestId('query-editor');
+    const regexInput = getByTestId(
+      selectors.pages.Dashboard.Settings.Variables.Edit.QueryVariable.queryOptionsRegExInputV2
+    );
+    const sortSelect = getByTestId(
+      selectors.pages.Dashboard.Settings.Variables.Edit.QueryVariable.queryOptionsSortSelectV2
+    );
+    const refreshSelect = getByTestId(
+      selectors.pages.Dashboard.Settings.Variables.Edit.QueryVariable.queryOptionsRefreshSelectV2
+    );
+
+    const multiSwitch = getByTestId(
+      selectors.pages.Dashboard.Settings.Variables.Edit.General.selectionOptionsMultiSwitch
+    );
+    const includeAllSwitch = getByTestId(
+      selectors.pages.Dashboard.Settings.Variables.Edit.General.selectionOptionsIncludeAllSwitch
+    );
+    const allValueInput = getByTestId(
+      selectors.pages.Dashboard.Settings.Variables.Edit.General.selectionOptionsCustomAllInput
+    );
+
+    expect(dataSourcePicker).toBeInTheDocument();
+    expect(dataSourcePicker).toHaveTextContent('Default Test Data Source');
+    expect(regexInput).toBeInTheDocument();
+    expect(regexInput).toHaveValue('.*');
+    expect(sortSelect).toBeInTheDocument();
+    expect(sortSelect).toHaveTextContent('Alphabetical (asc)');
+    expect(refreshSelect).toBeInTheDocument();
+    expect(getByRole(refreshSelect, 'radio', { name: 'On dashboard load' })).toBeChecked();
+    expect(multiSwitch).toBeInTheDocument();
+    expect(multiSwitch).toBeChecked();
+    expect(includeAllSwitch).toBeInTheDocument();
+    expect(includeAllSwitch).toBeChecked();
+    expect(allValueInput).toBeInTheDocument();
+    expect(allValueInput).toHaveValue('custom all value');
+  });
+
+  it('should call onDataSourceChange when changing the datasource', async () => {
+    const { getByTestId } = setup();
+    const dataSourcePicker = getByTestId(selectors.components.DataSourcePicker.inputV2);
+    await waitFor(async () => {
+      await userEvent.click(dataSourcePicker); // open the select
+      await userEvent.tab();
+    });
+    expect(mockOnDataSourceChange).toHaveBeenCalledTimes(1);
+    expect(mockOnDataSourceChange).toHaveBeenCalledWith(defaultDatasource);
+  });
+
+  it('should call onQueryChange when changing the query', async () => {
+    const { getByTestId } = setup();
+    const queryEditor = getByTestId(
+      selectors.pages.Dashboard.Settings.Variables.Edit.QueryVariable.queryOptionsQueryInput
+    );
+
+    await waitFor(async () => {
+      await userEvent.type(queryEditor, '-new');
```

Comment:
In my other PR, I found a guideline regarding the usage of `userEvent` https://github.com/grafana/grafana/blob/main/contribute/style-guides/testing.md#testing-user-interactions, based on the library guidelines is recommended to call the `userEvent.setup()` before rendering the component (_It applies workarounds to fake user interactions in the browser, since direct access to the UI layer isn't available_)


``` 
//Example from guideline

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

function setup(jsx: JSX.Element) {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  };
}

it('should render', async () => {
  const { user } = setup(<Button />);
  await user.click(screen.getByRole('button'));
});
```

Reply:
Good to know! Refactoring!